shared/util/helpers.ts

```diff
@@ -7,6 +7,18 @@ export function isDefined<T>(value: T | undefined): value is T {
     return value !== undefined
 }
 
+/**
+ * Returns true fi the value is defined and, if an array, contains at least
+ * one element.
+ *
+ * @param value The value to test.
+ */
+export function nonEmpty<T>(
+    value: T | T[] | null | undefined
+): value is T | T[] {
+    return !!value && !(Array.isArray(value) && value.length === 0)
```

Comment:
question:
wut sorcery is this? what is the `value is T | T[]` return type, and why `!!value`?

Reply:
`value is X` is a type assertion so that at the call site it will refine the type of the parameter inside the conditional.

`!!x` is equivalent to `Boolean(x)` - it negates the truthiness twice to make a boolean. Without it the expression returns bool or undefined, not bool (due to short circuiting of &&).